### PR TITLE
GH-593: Spec ctor not initializing inherited fields (specification update)

### PR DESCRIPTION
**Pull request by mor-n4**
See #593.

### CODE REVIEW

**Comment by mor-n4** [comment]
This is just the specification update (no actual changes, just added clarifications + major overhaul of the @Spec ctor section).

To be discussed if we want a warning for this case (then code from another PR will be used).

File: docs/org.eclipse.n4js.spec/chapters/05_classifiers/classifiers-part-02-members.adoc [hunk 1]
```diff
@@ -639,41 +639,138 @@ class B extends A {
...
-7.  The type of an additional member which match owned non-public fieldmust be subtype of the field’s type:
+3.  If a formal parameter is annotated with `@Spec`, the parameter’s type must be `pass:[~this]` or `pass:[~i~this]` (i.e.
+a use-site structural initializer field type of `this`, see <<structural-readWriteInit-field-typing>>).
+4.  A spec constructor will automatically initialize all its owned data fields, owned setters and data fields / setters
+from implemented interfaces that _are also part of the spec parameter's structural initializer field type_ with the data
+provided in the argument to the spec parameter.
+5.  Properties provided by an argument, but not defined in the parameter's structural initializer field type, are _not_
+used for initialization, even if a field of the same name exists in the class.
+6.  Fields explicitly added to the spec parameter, e.g. `@Spec spec: ~i~this with {name:string}`, are used for initialization
+if a non-public field of the same name exists in the class, either as an owned member of from an implemented interface.
```
Comment (mmews-n4) [inline docs/org.eclipse.n4js.spec/chapters/05_classifiers/classifiers-part-02-members.adoc]: Did I understand correctly?:

Fields explicitly added to the spec parameter, e.g. `@Spec spec: ~i~this with {name:string}`, are used for initialization if:
* a non-public field of the same name exists in the class in an implemented interface, and
* the type is a subtype of the class's field’s type.

File: docs/org.eclipse.n4js.spec/chapters/05_classifiers/classifiers-part-02-members.adoc [hunk 2]
```diff
@@ -639,41 +639,138 @@ class B extends A {
...
+The type of such an additional field must be a subtype of the class's field’s type.
```
Comment (mmews-n4) [inline docs/org.eclipse.n4js.spec/chapters/05_classifiers/classifiers-part-02-members.adoc]: > class's field’s type

(two different apostrophes)

Confusing. It means "the ((type of the field) of the class)"?

File: docs/org.eclipse.n4js.spec/chapters/05_classifiers/classifiers-part-02-members.adoc (hunk 1 above)
Comment (mor-n4) [reply docs/org.eclipse.n4js.spec/chapters/05_classifiers/classifiers-part-02-members.adoc]: yes

File: docs/org.eclipse.n4js.spec/chapters/05_classifiers/classifiers-part-02-members.adoc [hunk 3]
```diff
@@ -639,41 +639,151 @@ class B extends A {
...
 $\hspace{2em}\exists f \in ctor.owner.ownedFields \Rightarrow \tee s \subtype f $
-8.  `pass:[~i~this]` constructor ignores superfluous properties provided by an object literal.
-These ignored properties are _not_ used to set non-$\lenum{public}$ fields.
-9.  Since use-site structural initializer field types can be defined via public, non-static, non-optional writable fields, `pass:[~i~this]` constructor accepts those properties provided by an object literal which has the corresponding readble fields.
-These properties will be initialzed.
+6.  Even if the `@Spec` annotation is used, the super constructor must be invoked explicitly (as usual).
 
 --
 
+It follows from no. 4 above that
```
Comment (mmews-n4) [inline docs/org.eclipse.n4js.spec/chapters/05_classifiers/classifiers-part-02-members.adoc]: good clarification

**Comment by mor-n4** [comment]
very helpful feedback, thanks!

Merging...

**Pull request merged**